### ISSUE TITLE
Idiomatic mutex guards

### ISSUE THREAD

**Issue by ggwpez** [hyperledger-labs/go-perun#180]
Having two lines at the start of each function is quite ugly:  
```go
func (o Object) … {
    o.mtx.Lock()
    defer o.mtx.Unlock()
    …
```
I propose to add something like a `Guard` function which puts this in one line:
```go
func (o Object) … {
    defer pkgsync.Guard(&o.mtx)()
    …
```
This can be extended for `RWMutex` with `GuardR` and `GuardRW`. The implementation of each function are two lines.

**Comment by manoranjith** [comment]
@ggwpez What would be the implementation of Gaurd ?

If we put the two lines inside the function, will both (lock & unlock) not be deferred ? 
```
func Gaurd(mtx *sync.Mutex) {
    mtx.Lock()
    mtx.Unlock()
}

**Comment by ggwpez** [comment]
> @ggwpez What would be the implementation of Gaurd ?
> 
> If we put the two lines inside the function, will both (lock & unlock) not be deferred ?
> 
> ```
> func Gaurd(mtx *sync.Mutex) {
>     mtx.Lock()
>     mtx.Unlock()
> }
> ```

For example:  
```go
func Guard(mtx *sync.Mutex) context.CancelFunc {
	mtx.Lock()
	return mtx.Unlock
}
```
The use of `context.CancelFunc` is not mandatory and can be replaced by `func()`.  
@RmbRT suggested it since it may raise linter warnings when it is ignored by a malformed `defer Guard(&mtx)` call.  
@manoranjith

**Comment by manoranjith** [comment]
Yeah, I agree using `CancelFunc` is an nice trick to take advantage of linter.

But, I think the our function might not satisfy the properties of `CancelFunc`.
[`CancelFunc`](https://pkg.go.dev/context#CancelFunc) is safe for concurrent use and does not error on multiple invocations.
But, for `Unlock()` [invoking it for the second time, (when it is already unlocked), will result in runtime error.](https://pkg.go.dev/sync#Mutex.Unlock)

In either case, I see user should be careful
1. `func()`: they should not forget to call it.
2. `CancelFun`: they must call it only once.

To, me 1 seems like a cleaner approach.

**Comment by RmbRT** [comment]
Since the function would only be used as `defer Guard(&mutex)()` any way, it's basically impossible to call the unlock method multiple times. However, it's much easier to accidentally forget the last `()`, so I think the Cancel function should be used. If you don't use the Guard function in the defer statement, you are better off using the normal `lock(); defer unlock();` pattern any way.

**Comment by matthiasgeihs** [comment]
I think the issue title is misleading. *Idiomatic* means *using, containing, or denoting expressions that are natural to a native speaker* (according to Google Search).

And this also points to the issue I have with this proposal. I think it is exactly non-idiomatic and therefore problematic. Usage will be unclear and unintuitive to other golang programmers, which potentially leads to misusage and bugs.

If you feel the language could be improved, I think the right place would be https://github.com/golang/proposal.

**Comment by ggwpez** [comment]
Ok then lets leave it as is.
